memstat.1
.\"  This page Copyright (C) 1997 Joshua Yelon <jyelon@uiuc.edu>
.\"  Distribution subject to the terms of the GPL.
.\"  minor fixes by Bernd Eckenfels <ecki@debian.org> 1998-04-18
.\"  minor fixes by Bernd Eckenfels <ecki@debian.org> 1998-11-01
.\"  significan rewrite by Michael Meskes <meskes@debian.org> 2009-04-06
.TH MEMSTAT 1 "01 November 1998 " "Debian" "Linux Programmer's Manual"
.SH NAME
memstat \- Identify what's using up virtual memory.
.SH SYNOPSIS
.ft B
.B memstat
.RB [ "\-w" ] [ "\-p PID" ]
.br
.SH DESCRIPTION
\fBmemstat\fP lists all accessible processes, executables, and shared libraries
that are using up virtual memory. To get a complete list \fBmemstat\fP has to
be run as root to be able to access the data of all running processes.

First, the processes are listed.  An amount of memory is shown along
with a process ID and the name of the executable which the process is
running.  The amount of memory shown does not include shared memory:
it only includes memory which is private to that process.  So, if a
process is using a shared library like libc, the memory used to hold
that library is not included.  The memory used to hold the
executable's text-segment is also not included, since that too is
shareable.

After the processes, the shared objects are listed.  The amount of
memory is shown along with the filename of the shared object, followed by a
list of the processes using the shared object. The memory is listed as the
total amount of memory allocated to this object throughout the whole namespace.
In brackets also the amount that is really shared is listed.

Finally, a grand total is shown.  Note that this program shows the
amount of \fBvirtual\fP (not real) memory used by the various items.

\fBmemstat\fP gets its input from the /proc filesystem.  This must be
compiled into your kernel and mounted for \fBmemstat\fP to work.  The
pathnames shown next to the shared objects are determined by scanning
the disk.  \fBmemstat\fP uses a configuration file, /etc/memstat.conf,
to determine which directories to scan.  This file should include all
the major bin and lib directories in your system, as well as the /dev
directory.  If you run an executable which is not in one of these
directories, it will be listed by memstat as ``[0dev]:<inode>''.

.SS Options
The \fB-w\fP switch causes a wide printout: lines are not truncated at
80 columns.
.PP
The \fB-p\fP switch causes \fBmemstat\fP to only print data gathered from
looking at the process with the gicen PID.
.SH NOTES
These reports are intended to help identify programs that are using an
excessive amount of memory, and to reduce overall memory waste.
.PP
.SH FILES
.ta
.nf
/etc/memstat.conf
/proc/*/maps
.fi

.SH "SEE ALSO"
ps(1), top(1), free(1), vmstat(8), lsof(8), /usr/share/doc/memstat/memstat-tutorial.txt.gz
.PP
.SH BUGS

\fBmemstat\fP ignores all devices that just map main memory, though this may
cause \fBmemstat\fP to ignore some memory usage.

Memory used by the kernel itself is not listed.
.SH AUTHOR
Originally written by Joshua Yelon <jyelon@uiuc.edu> and patched by Bernd Eckenfels
<ecki@debian.org>. Taken over and rewritten by Michael Meskes <meskes@debian.org>.


debian/changelog
memstat (0.8) unstable; urgency=low

  * Fixed handling of dev attribute. The old version failed to work under some
    circumstances, closes: #531757
  * Re-added output of PIDs that was somehow lost, closes: #531682

 -- Michael Meskes <meskes@debian.org>  Mon, 08 Jun 2009 14:15:57 +0200

memstat (0.7) unstable; urgency=low

  * If a process is stopped while processing, memstat no longer stops, but
    instead prints out an error and continues working, closes: #531435

 -- Michael Meskes <meskes@debian.org>  Tue, 02 Jun 2009 15:13:00 +0200

memstat (0.6) unstable; urgency=low

  * Corrected path in reference to memstat-tutorial.txt.gz, closes: #502844
  * Made program read name of executable instead of guessing it from the memory
    information, closes: #503026
  * Added option to only show information for one process.
  * Do not list memory mapping devices, closes: #47260, #58290, #108510
  * List all other device mappings, closes: #496778
  * Count double mapped pages twice, closes: #496777
  * Completely redid the whol calculation part.
  * Bumped Standards-Version to 3.8.1, no changes needed.

 -- Michael Meskes <meskes@debian.org>  Mon, 06 Apr 2009 16:11:19 +0200

memstat (0.5) unstable; urgency=low

  * New maintainer, closes: #485998
  * Aknowledged NMU, closes: #322749
  * Fixed compiler warnings by applying patch, closes: #356024
  * Fixed wrong pathname problem by applying patch, closes: #294635, #279117
  * Bumped Standards-Version to 3.8.0.
  * Bumped debhelper compat version to 7.
  * Simplified rules file.

 -- Michael Meskes <meskes@debian.org>  Fri, 13 Jun 2008 16:54:39 +0200

memstat (0.4.0.0.1) unstable; urgency=low

  * Non-maintainer upload (take two.. where did the first one go?)
  * Complete /usr/doc transition. Closes: #322749

 -- Joey Hess <joeyh@debian.org>  Tue, 10 Jan 2006 01:28:34 -0500

memstat (0.4) unstable; urgency=high

  * make fscanf buffer bigger and more dynamic (Closes: Bug#5544)
  * less verbose postinst (Closes: Bug#97400)

 -- Bernd Eckenfels <ecki@debian.org>  Fri, 12 Jul 2002 05:37:51 +0200

memstat (0.3) unstable; urgency=low

  * catch null pointer in reallof (Closes Bug: #108767)
  * new policy version (3.5.6) and use of strip/debug switches
  * includes Build-Depends now
  * lintian clean (added Priority, fixed spelling, added -isp)

 -- Bernd Eckenfels <ecki@debian.org>  Thu, 23 May 2002 04:26:51 +0200

memstat (0.2) unstable; urgency=low

  * support for path in maps files (#23099, #31043, #49855, #50700, #22674)
  * added libc5 compat dirs (#34326)
  * started to maintain the package upstream
  * no compiler warnings and lintian clean (support for /usr/share)
  * new standards version

 -- Bernd Eckenfels <ecki@debian.org>  Sat, 20 Nov 1999 18:40:23 +0100

memstat (0.1-3) unstable; urgency=low

  * removed double /usr/local/bin in conffile (Bug#22674)
  * moved man page to section 1 and meminfo -> memstat (Bug#23096)
  * added debian/conffiles to .deb archives control dir (Bug#27818)

 -- Bernd Eckenfels <ecki@debian.org>  Sun,  1 Nov 1998 09:08:05 +0100

memstat (0.1-2) unstable; urgency=low

  * modified manpage to show new 'unknown' format and fix Bug#21360

 -- Bernd Eckenfels <ecki@debian.org>  Sat, 18 Apr 1998 21:58:23 +0200


memstat (0.1-1) unstable; urgency=low

  * changed the output from unknown files to "[<dev>]:<inode>"
  * Initial Release

 -- Bernd Eckenfels <ecki@debian.org>  Mon,  6 Apr 1998 22:56:41 +0200

Local variables:
mode: debian-changelog
End:


memstat.c
/*
 * This software copyright 1997 Joshua M. Yelon.
 * Debian Maintainer and new Features:
 *   copyright 1999 Bernd Eckenfels, Germany, ecki@debian.org 
 *   (see debian/changelog)
 *   copyright 2008-2009 Michael Meskes, meskes@debian.org
 *
 * Distribution subject to the terms of the GPL.
 *
 */

#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <dirent.h>
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>

/* blacklist devices that just map physical memory */
char *blacklist[] = { "/dev/mem",
    "/dev/dri/"
};

int wide = 0;
int only_pid = 0;
int needinode = 0;

typedef int (*qcmp) (const void *, const void *);

typedef struct mapping *mapping;

struct mapping {
    unsigned int fs, inode, pid;
    unsigned long offs, lo, hi;
    char *path;
    int valid;
    int shared;
};

mapping maptab_data;
size_t maptab_fill;
size_t maptab_size;

static void maptab_expand(void)
{
    size_t bytes;

    bytes = (maptab_size * 2 + 100) * sizeof(struct mapping);
    maptab_data = (mapping) realloc(maptab_data, bytes);
    if (maptab_data == NULL) {
	fprintf(stderr, "Error: realloc returned null, possibly out of memory? Exiting.\n");
	exit(1);
    }
    maptab_size = maptab_size * 2 + 100;
}

static void read_proc(void)
{
    unsigned int nread, pid;
    unsigned long inode, lo, hi, offs;
    char *p, major[8], minor[8], buff[PATH_MAX + 300], *path, perm[4];
    DIR *d;
    struct dirent *ent;
    FILE *f;

    mapping m;
    d = opendir("/proc");
    if (d == 0) {
	perror("/proc");
	exit(1);
    }
    while (1) {
	ent = readdir(d);
	if (ent == NULL)
	    break;
	pid = strtol(ent->d_name, NULL, 10);
	if (errno != 0) {
		perror("strtol");
		exit(1);
	}
	if (pid == 0 || (only_pid != 0 && pid != only_pid))
	    continue;
	sprintf(buff, "/proc/%d/maps", pid);
	f = fopen(buff, "r");
	if (f == NULL)
	    continue;
	while (fgets(buff, sizeof(buff), f)) {
	    p = strchr(buff, '-');
	    if (p)
		*p = ' ';
	    p = strchr(buff, ':');
	    if (p)
		*p = ' ';
	    path = NULL;
	    if ((strlen(buff) == 10) && (strcmp(buff, " (deleted)") == 0))
		continue;
	    nread = sscanf(buff, "%lx %lx %4s %lx %s %s %lu %as", &lo, &hi, perm, &offs, major, minor, &inode, &path);
	    if (nread < 7) {
		fprintf(stderr, "I don't recognize format of /proc/%d/maps. (nread=%d)\n", pid, nread);
		exit(1);
	    }
	    if (maptab_fill == maptab_size)
		maptab_expand();
	    m = maptab_data + maptab_fill;
	    /* under some circumstances strtol returns errno still set to an old value */
	    /* let's play it safe here and set errno to 0 to not error out */
	    errno = 0;
	    m->fs = (dev_t)(strtol(major, NULL, 16) * 256 + strtol(minor, NULL, 16));
	    if (errno != 0) {
		perror("strtol");
                exit(1);
            }
	    m->inode = inode;
	    m->offs = offs;
	    m->pid = pid;
	    m->lo = lo;
	    m->hi = hi;
	    m->valid = 1;
	    if ((nread == 8) && path && path[0]) {
		int i;

		m->path = path;
		for (i = 0; i < sizeof(blacklist) / sizeof(blacklist[0]); i++) {
		    if (!strncmp(path, blacklist[i], strlen(blacklist[i])))
			m->valid = 0;
		}
	    } else {
		sprintf(buff, "[%s:%s]:%lu", major, minor, inode);
		m->path = strdup(buff);
		needinode = 1;
	    }
	    /* mark segment as either shared or private */
	    if (perm[3] == 's' ||	/* declared as shared */
		perm[2] == 'x')		/* executables are shared */
		m->shared = 1;
	    else
		m->shared = 0;
	    maptab_fill++;
	}
	fclose(f);
    }
    closedir(d);
}

static int sort_by_pid(mapping m1, mapping m2)
{
    int delta;

    delta = m1->pid - m2->pid;
    return delta;
}

static int sort_by_inode(mapping m1, mapping m2)
{
    int delta;

    delta = m1->fs - m2->fs;
    if (delta)
	return delta;
    delta = m1->inode - m2->inode;
    if (delta)
	return delta;
    delta = m1->shared - m2->shared;
    if (delta)
	return delta;
    delta = m1->offs - m2->offs;
    return delta;
}

static void register_path(unsigned int fs, unsigned int inode, char *path, int valid)
{
    size_t i;
    mapping m;
    char *copy = NULL;

    for (i = 0; i < maptab_fill; i++) {
	m = maptab_data + i;
	if ((m->path) && (m->path[0] == '[') && (m->fs == fs)
	    && (m->inode == inode)) {
	    if (copy == NULL)
		copy = strdup(path);
	    m->path = copy;
	    m->valid = valid;
	}
    }
}

static void scan_directory(char *dir)
{
    DIR *d;
    struct dirent *ent;
    struct stat info;
    int ok;
    char full[8192];

    d = opendir(dir);
    if (d == NULL)
	return;
    while ((ent = readdir(d)) != NULL) {
	sprintf(full, "%s/%s", dir, ent->d_name);
	ok = stat(full, &info);
	if (ok >= 0)
	    register_path(info.st_dev, info.st_ino, full, S_ISREG(info.st_mode));
    }
    closedir(d);
}

static void scan_fpath(const char *fn)
{
    char *p;
    FILE *f = fopen(fn, "r");
    char buff[1024];

    if (f == NULL) {
	fprintf(stderr, "Cannot open /etc/memstat.conf\n");
	exit(1);
    }
    while (fgets(buff, 1023, f)) {
	if (buff[0] == '#')
	    continue;
	p = strchr(buff, '\n');
	if (p)
	    *p = 0;
	scan_directory(buff);
    }
    fclose(f);
}

static void printline(char *str)
{
    if (!wide && strlen(str) > 79) {
	str[76] = '.';
	str[77] = '.';
	str[78] = '.';
	str[79] = 0;
    }
    printf("%s\n", str);
}

static void summarize_usage(void)
{
    char buffer[8192];
    unsigned int i, fs, inode, pid, scan;
    unsigned long offs, grand, lo = 0, hi = 0, total, sharedtotal, sharedgrand;
    mapping m;
    char *exe;

    grand = sharedgrand = 0;
    qsort(maptab_data, maptab_fill, sizeof(struct mapping), (qcmp) sort_by_pid);
    for (offs = 0; offs < maptab_fill; offs = scan) {
	char linkname[PATH_MAX], filename[PATH_MAX];
	ssize_t len;
	int deleted = 0;

	pid = maptab_data[offs].pid;
	sprintf(filename, "/proc/%d/exe", pid);
	if ((len = readlink(filename, linkname, PATH_MAX)) == -1) {
	    fprintf(stderr, "Cannot read link information for %s\n", filename);
	    deleted = 1;
	}
	linkname[len] = '\0';
	total = 0;
	for (scan = offs; scan < maptab_fill; scan++) {
	    m = maptab_data + scan;
	    if (m->pid != pid)
		break;
	    /* if fs and/or inode are set, memory belongs to the process listed */
	    /* [vsyscall] is in kernel space */
	    if (!deleted && !m->fs && !m->inode && strcmp(m->path, "[vsyscall]"))
		total += (m->hi - m->lo);
	}
	if (!deleted) {
		sprintf(buffer, "%7ldk: PID %5d (%s)", total / 1024, pid, linkname);
		printline(buffer);
		grand += total;
	}
    }

    qsort(maptab_data, maptab_fill, sizeof(struct mapping), (qcmp) sort_by_inode);
    for (offs = 0; offs < maptab_fill; offs = scan) {
	m = maptab_data + offs;
	if (m->valid == 0) {
	    scan = offs + 1;
	    continue;
	}
	fs = m->fs;
	inode = m->inode;
	exe = m->path;
	if ((fs == 0) && (inode == 0)) {
	    scan = offs + 1;
	    continue;
	}
	sharedtotal = total = hi = 0;
	for (scan = offs; scan < maptab_fill; scan++) {
	    m = maptab_data + scan;
	    if (fs != m->fs || inode != m->inode) {	/* new file */
		break;
	    } else if (m->shared) {
#if 0
		if (hi == 0) {	/* first shared segment */
		    lo = m->offs;
		    hi = m->offs + m->hi - m->lo;
		} else if (m->offs != lo) {	/* new unconnected segment */
		    total += (hi - lo);
		    lo = m->offs;
		    hi = m->offs + m->hi - m->lo;
		} else if (m->offs + m->hi - m->lo > hi) {	/* longer than the one we had earlier */
		    hi = m->offs + m->hi - m->lo;
		}
#else
		if (hi == 0) {
		    lo = m->offs;
		    hi = m->offs + m->hi - m->lo;
		} else if (hi != m->offs + m->hi - m->lo) {
		    sharedtotal += (hi - lo);
		    lo = m->offs;
		    hi = m->offs + m->hi - m->lo;
		}
#endif
	    } else
		total += (m->hi - m->lo);
	}
	/* do we have a segment left that hasn't been counted? */
	if (hi != 0)
	    sharedtotal += (hi - lo);
	sprintf(buffer, "%7ldk(%7ldk): %s", (total + sharedtotal) / 1024, sharedtotal / 1024, exe);

	pid = 0;
	for (i = offs; i < scan; i++) {
	    m = maptab_data + i;
	    if (m->pid != pid) {
		pid = m->pid;
		sprintf(buffer + strlen(buffer), " %d", pid);
	    }
	}
	printline(buffer);
	grand += total;
	sharedgrand += sharedtotal;
    }
    printf("--------\n");
    printf("%7ldk (%7ldk)\n", (grand + sharedgrand) / 1024, sharedgrand / 1024);
}

static void usage(char *prog)
{
    fprintf(stderr, "usage: %s [-w] [-p pid]\n", prog);
    exit(1);
}

int main(int argc, char **argv)
{
    char *prog = argv[0];
    int opt;

    while ((opt = getopt(argc, argv, "wp:")) != -1) {
	switch (opt) {
	case 'w':
	    wide = 1;
	    break;
	case 'p':
	    only_pid = atoi(optarg);
	    break;
	default:
	    usage(prog);
	}
    }

    read_proc();
    if (needinode)
	scan_fpath("/etc/memstat.conf");
    summarize_usage();
    return (0);
}
